src/java.base/share/classes/java/security/Provider.java

```diff
@@ -749,196 +1464,195 @@ private void readObject(ObjectInputStream in)
-        } else {
-            return true;
+    /*
+     * Enum to determine if changes to the Properties map must be applied by the
+     * caller (UPDATE) or skipped (SKIP).
+     *
+     * If a change does not concern a ServicesMap, UPDATE is returned. An
+     * example of this is when adding, modifying or removing an entry that is
+     * not a service, alias or attribute.
+     *
+     * If the change concerns a ServicesMap, SKIP is returned. The change may
+     * have been applied internally or ignored due to an error. In the former
+     * case, Properties map entries are synchronized. In the latter, Properties
+     * map entries are not modified.
+     */
+    private enum PropertiesMapAction {
+        UPDATE,
+        SKIP
+    }
+
+    private PropertiesMapAction doLegacyOp(ServicesMap servicesMap, Object key,
+            Object value, Object oldValue, OPType opType) {
+        if (key instanceof String ks) {
+            if (ks.startsWith("Provider.")) {
+                // Ignore provider related updates.
+                return PropertiesMapAction.SKIP;
+            }
+            if (value instanceof String vs) {
+                return parseLegacy(servicesMap, ks, vs, opType);
+            } else if (value != null && oldValue instanceof String oldValueS &&
+                    opType == OPType.ADD) {
+                // An entry in the Properties map potentially concerning the
+                // ServicesMap is about to be replaced by one that does not.
+                // From the ServicesMap point of view, this could be equivalent
+                // to a removal. In any case, let the caller proceed with the
+                // Properties map update.
+                parseLegacy(servicesMap, ks, oldValueS, OPType.REMOVE);
+            }
         }
+        // The change does not concern a ServicesMap.
+        return PropertiesMapAction.UPDATE;
     }
 
     /**
      * Copies all the mappings from the specified Map to this provider.
      */
     private void implPutAll(Map<?,?> t) {
+        // For service readers to see this change as atomic, add the elements in
+        // a local thin copy of the ServicesMap and then publish it.
+        ServicesMap servicesMapCopy = new ServicesMap(servicesMap);
         for (Map.Entry<?,?> e : t.entrySet()) {
-            implPut(e.getKey(), e.getValue());
+            implPut(servicesMapCopy, e.getKey(), e.getValue());
         }
+        servicesMap = servicesMapCopy;
     }
 
-    private Object implRemove(Object key) {
-        if (!checkLegacy(key)) return null;
+    private void implClear() {
+        servicesMap.clear();
+        prngAlgos.clear();
+        super.clear();
+        putId();
+    }
 
-        Object o = super.remove(key);
-        if (o instanceof String so && key instanceof String sk) {
-            parseLegacy(sk, so, OPType.REMOVE);
-        }
-        return o;
+    private Object implRemove(Object key) {
+        Object oldValue = super.get(key);
```

Comment:
`oldValue` isn't really old value, is it? Naming it `oldValue` and pass it to `doLegacyOp()` as `value` is very confusing.

Reply:
Yes, I see why it could be confusing. We named it `old` thinking from the returned value's point of view (when there is a removal, we have to return the old/previous value). Perhaps we should rename it to just `value`.